src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,70 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+-- XXX We can chain this with another parser to propagate the error and act
+-- accordingly.
```

Comment:
This comment can be removed. If it is informative only comment then XXX must not be used.